### ISSUE TITLE
incremental mask misbehaves for DEScrypt & LM opencl formats

### ISSUE THREAD

**Issue by AlekseyCherepanov** [openwall/john#5084]
I noticed a problem with `--mask='[a-z]?a' --min-length=1 --max-length=5` with `descrypt-opencl`. Just `?a` works as expected. `descrypt` seems ok.

```
$ echo AAAAAAAAAAAAA > t.pw
$ ./john/run/john --format=descrypt-opencl t.pw --mask='[a-z]?a' --min-length=1 --max-length=5 --session=s/s2
Device 1: Hawaii [AMD Radeon (TM) R9 390 Series]
Using default input encoding: UTF-8
Loaded 1 password hash (descrypt-opencl, traditional crypt(3) [DES OpenCL])
Note: descrypt-opencl format can't currently increment length from 1, using 2 instead
Build log: /tmp/OCL194688T2.cl:208:23: warning: unknown attribute 'max_constant_size' ignored
                                   __attribute__((max_constant_size((ACTIVE_PLACEHOLDER * 32 * ITER_COUNT))))
                                                  ^
1 warning generated.

LWS=128 GWS=262144 x95
Press 'q' or Ctrl-C to abort, 'h' for help, almost any other key for status
0g 0:00:00:00  (2) 0g/s 0p/s 0c/s 0C/s

0g 0:00:00:03 31.75% (2) (ETA: 23:40:51) 0g/s 210783Kp/s 210783Kc/s 210783KC/s maH9vU..maH9vU
0g 0:00:00:09 49.99% (5) (ETA: 23:41:00) 0g/s 216443Kp/s 216443Kc/s 216443KC/s gac>|U..gac>|U
0g 0:00:00:19 DONE (5) (2022-02-20 23:41) 0g/s 218021Kp/s 218021Kc/s 218021KC/s map~}..map~}
Warning: incremental mask started at length 2 - try the CPU format for shorter lengths.
Session completed. 
```

Notice `31.75% (2) [...] maH9vU..maH9vU`. It has `(2)` but candidates shown as current have length 6 (while `--max-length=5`). It is a problem.

There is `Note: [...]` about length. `--min-length=2` removes it, but does not solve the problem.

The problem is not cosmetic. It cannot crack 2 chars long password:
```
$ echo AAQg8DP4VZOHs > t.pw
$ ./john/run/john --format=descrypt-opencl t.pw --mask='[a-z]?a' --min-length=2 --max-length=5
[...]
0g 0:00:00:19 DONE (5) (2022-02-20 23:54) 0g/s 219256Kp/s 219256Kc/s 219256KC/s map~}..map~}
Session completed. 

$ ./john/run/john --format=descrypt-opencl t.pw --mask='[a-z]?a'
[...]
ab               (?)     
1g 0:00:00:00  4.347g/s 10739p/s 10739c/s 10739C/s ab..aa
[...]
```

**Comment by magnumripper** [comment]
Seems to be a driver bug - it works fine with nvidia:
```
$ echo AAQg8DP4VZOHs > t.pw
$ ../run/john --format=descrypt-opencl t.pw --mask='[a-z]?a' --min-length=2 --max-length=5
Device 1: NVIDIA GeForce RTX 2080 Ti
Using default input encoding: UTF-8
Loaded 1 password hash (descrypt-opencl, traditional crypt(3) [DES OpenCL])
LWS=128 GWS=131072 x2470
Press 'q' or Ctrl-C to abort, 'h' for help, almost any other key for status
0g 0:00:00:00  (5) 0g/s 0p/s 0c/s 0C/s Dev#1:31°C util:12%
ab               (?)     
1g 0:00:00:00 DONE (5) (2022-02-21 14:49) 4.761g/s 1541Mp/s 1541Mc/s 1541MC/s Dev#1:33°C util:12% ab..aa
No remaining hashes
Use the "--show" option to display all of the cracked passwords reliably
Session completed. 
```
It also works with `--min-len=1` (it starts at length 2 after that warning, and does crack this hash).

**Comment by magnumripper** [comment]
> Notice `31.75% (2) [...] maH9vU..maH9vU`. It has `(2)` but candidates shown as current have length 6 (while `--max-length=5`). It is a problem.

That is indeed weird. Same there, I can't see that problem with the nvidia.

**Comment by magnumripper** [comment]
> Seems to be a driver bug - it works fine with nvidia:

It could also be a bug of ours that somehow surfaces with AMD but not nvidia. Not sure what it would be.

**Comment by magnumripper** [comment]
@AlekseyCherepanov please try similar tests with the LM-opencl format - it has quirks similar to DEScrypt.

**Comment by magnumripper** [comment]
Formats with internal mask are handled specially with incrementing length mask, it goes like this: We fill candidates for `[a-z]`. Once we switch to length 2 we need to call `reset()` to set the format up for `[a-z]?a` (if the **internal** part of it changed - so this only happens once or twice with the shortest lengths). Before that, we must ensure that the current (incomplete) batch is processed so we call `crk_process_buffer()` to do so.

I could not get that to work for the DEScrypt and LM formats, so they refuse to do the above and instead react with eg. that "Note: descrypt-opencl format can't currently increment length from 1, using 2 instead" message and the longer parts are on CPU side. In this case the internal part is fixed as `[a-z]?a` (on the nvidia) and the host code adds more letters after it.

I see my nvidia picked a larger internal part `x2470` whereas your card ended up as `x95`. So let's enforce the smaller with nvidia:
```
$ echo AAQg8DP4VZOHs > t.pw
$ ../run/john --format=descrypt-opencl t.pw --mask='[a-z]?a' --min-length=1 --max-length=5 --mask-internal-target=95
Device 1: NVIDIA GeForce RTX 2080 Ti
Using default input encoding: UTF-8
Loaded 1 password hash (descrypt-opencl, traditional crypt(3) [DES OpenCL])
Note: descrypt-opencl format can't currently increment length from 1, using 2 instead
LWS=128 GWS=1048576 x95
Press 'q' or Ctrl-C to abort, 'h' for help, almost any other key for status
0g 0:00:00:00  (2) 0g/s 0p/s 0c/s 0C/s Dev#1:34°C util:25%
0g 0:00:00:02 49.69% (5) (ETA: 15:35:42) 0g/s 757938Kp/s 757938Kc/s 757938KC/s Dev#1:37°C util:77% uae'}U..uae'}U
0g 0:00:00:05 N/A 0g/s 776999Kp/s 776999Kc/s 776999KC/s Dev#1:39°C util:77% oaiF>..oaiF>
Warning: incremental mask started at length 2 - try the CPU format for shorter lengths.
Session completed. 
```
Indeed, I can reproduce.

If I force it to just 26 (taking care of the very first character), it again works (note that this also makes `--min-length=1` work):
```
$ ../run/john --format=descrypt-opencl t.pw --mask='[a-z]?a' --min-length=1 --max-length=5 --mask-internal-target=26
Device 1: NVIDIA GeForce RTX 2080 Ti
Using default input encoding: UTF-8
Loaded 1 password hash (descrypt-opencl, traditional crypt(3) [DES OpenCL])
LWS=128 GWS=1048576 x26
Press 'q' or Ctrl-C to abort, 'h' for help, almost any other key for status
0g 0:00:00:00  (5) 0g/s 0p/s 0c/s 0C/s Dev#1:36°C util:37%
ab               (?)     
1g 0:00:00:00 DONE (5) (2022-02-21 15:36) 5.555g/s 151460Kp/s 151460Kc/s 151460KC/s Dev#1:37°C util:18% ab..a
No remaining hashes
Use the "--show" option to display all of the cracked passwords reliably
Session completed. 
```
I bet if you force your card to `--mask-internal-target=2470` it will work for you too.

The problem is likely in `mask.c` and probably with code that depends on `format_cannot_reset` variable.

**Comment by magnumripper** [comment]
Same bug affects LM-opencl similarly.

**Comment by magnumripper** [comment]
This seems to be a complete fix, and isn't particularly strange considering surrounding code:
```diff
diff --git a/src/mask.c b/src/mask.c
index 6f11205c4..8dd730e9c 100644
--- a/src/mask.c
+++ b/src/mask.c
@@ -1612,8 +1612,9 @@ static MAYBE_INLINE char* mask_utf8_to_cp(const char *in)
 
 #define init_key(ps)							\
 	while (ps < MAX_NUM_MASK_PLHDR) {				\
-		template_key[ranges(ps).pos + ranges(ps).offset] =	\
-		ranges(ps).chars[ranges(ps).iter];			\
+		if (ranges(ps).pos + ranges(ps).offset < mask_cur_len) \
+			template_key[ranges(ps).pos + ranges(ps).offset] = \
+				ranges(ps).chars[ranges(ps).iter]; \
 		ps = ranges(ps).next;					\
 	}
 
```
It doesn't just empirically happen to fix the described issue, but also fixes fundamental problems clearly seen when debugging.

However, I don't fully understand it yet, so more testing might be needed. I did test hybrid mask mode with this in place and it appears unaffected, which is kinda strange - I thought it would break unless a little more logic added.

**Comment by magnumripper** [comment]
Some more testing shows more logic is indeed needed:
```diff
diff --git a/src/mask.c b/src/mask.c
index 6f11205c4..01e11834f 100644
--- a/src/mask.c
+++ b/src/mask.c
@@ -1612,8 +1612,9 @@ static MAYBE_INLINE char* mask_utf8_to_cp(const char *in)
 
 #define init_key(ps)							\
 	while (ps < MAX_NUM_MASK_PLHDR) {				\
-		template_key[ranges(ps).pos + ranges(ps).offset] =	\
-		ranges(ps).chars[ranges(ps).iter];			\
+		if (!mask_increments_len || ranges(ps).pos + ranges(ps).offset < mask_cur_len) \
+			template_key[ranges(ps).pos + ranges(ps).offset] = \
+				ranges(ps).chars[ranges(ps).iter]; \
 		ps = ranges(ps).next;					\
 	}
 
```
This feels closer to something I'd consider writing a PR for. The remaining thing nagging me is why no other format needs this.

EDIT: Nope, this is not complete yet.

**Comment by AlekseyCherepanov** [comment]
Sorry for huge delay on my side!

> I bet if you force your card to `--mask-internal-target=2470` it will work for you too.

Yes, it works.

> Same bug affects LM-opencl similarly.

Yes, it happens for me too.

The fix works.

I found other problems:
- `--mask-internal-target=0` gives segfault for `lm-opencl`
- `--mask='bbb?a' --min-length=1 --max-length=2` tells `using 4 instead`

Also `--format=descrypt-opencl --mask='bb?a?a?a?a' --min-length=1 --max-length=2 --mask-internal-target=69833729609375` caused my gpu to hang. So I'll file separate issues a bit later.

**Comment by magnumripper** [comment]
> * `--mask-internal-target=0` gives segfault for `lm-opencl`

> Also `--format=descrypt-opencl --mask='bb?a?a?a?a' --min-length=1 --max-length=2 --mask-internal-target=69833729609375` caused my gpu to hang. So I'll file separate issues a bit later.

We should bail with error for anything less than one, or larger than `pow(256, 4)`. Does your GPU hang if you lower that to `--mask-internal-target=81450625`?

**EDIT:** while `pow(256, 4)` is the theoretical max. an internal mask could end up as, it's obviously too large. We need to include GWS in the calculation and end up at max. 4G keys per crypt, I think a max. of `pow(256, 2)` is likely sufficient for now - that times a GWS of 64K is right at our ceiling.

> * `--mask='bbb?a' --min-length=1 --max-length=2` tells `using 4 instead`

So the problem this is that 4 is larger than `--max-length=2`. We need a couple lines of code handling that situation.

Please open two separate issues, thanks!

**Comment by magnumripper** [comment]
> We should bail with error for anything less than one

It turns out we should allow `--mask-internal-target=0`, it should simply turn off internal mask. It only worked incorrectly for DESCrypt and LM, bug fix coming.

Also, handling values larger than 32-bit can take is a more general problem with our options parser. They will be truncated but that happens with just any option that takes a number as a argument so I wont address it now.

> Please open two separate issues, thanks!

Never mind, unless you see problems after #5103.

**Comment by AlekseyCherepanov** [comment]
`--mask-internal-target=0` does not crash after #5103.

I played with masks like `--mask='bbbbbbb?a'` and different min/max lengths. It seems robust.

High values `--mask-internal-target=` still causes hangs. I'll file separate issue.

Thanks!

**Comment by magnumripper** [comment]
> High values `--mask-internal-target=` still causes hangs. I'll file separate issue.

I'm reluctant to change that as it's merely a "you get what you asked for" issue. Crazy big numbers will obviously tax the GPU just like it would for GWS or several other parameters.
